presto-main/src/main/java/io/prestosql/cost/TableScanStatsRule.java

```diff
@@ -73,11 +82,20 @@ public TableScanStatsRule(Metadata metadata, StatsNormalizer normalizer)
                     .orElse(SymbolStatsEstimate.unknown());
             outputSymbolStats.put(symbol, symbolStatistics);
         }
-
-        return Optional.of(PlanNodeStatsEstimate.builder()
-                .setOutputRowCount(tableStatistics.getRowCount().getValue())
-                .addSymbolStatistics(outputSymbolStats)
-                .build());
+        PlanNodeStatsEstimate estimate = PlanNodeStatsEstimate.builder()
+                                                              .setOutputRowCount(tableStatistics.getRowCount().getValue())
+                                                              .addSymbolStatistics(outputSymbolStats)
+                                                              .build();
+        // Allow the connector delegate filtering statistics estimation of the pushed-down predicate back to the engine:
+        TupleDomain<ColumnHandle> unestimatedPredicate = tableStatistics.getUnestimatedPredicate();
+        if (!unestimatedPredicate.isAll()) {
+            // If the connector don't support statistics estimation for pushed-down predicate, we use FilterStatsCalculator
+            // to heuristically update the PlanNodeStatsEstimate (similarly to how it's done by FilterStatsRule).
+            Map<ColumnHandle, Symbol> assignments = ImmutableBiMap.copyOf(node.getAssignments()).inverse();
+            Expression predicate = domainTranslator.toPredicate(unestimatedPredicate.simplify().transform(assignments::get));
```

Comment:
I worry that accuracy of stats can be greatly degraded by this operation.

Maybe instead when pushing down predicates, the engine would also pushdown it's current estimates for pushed down expressions (e.g: `ColumnStatistics` after predicate is applied on raw table stats).

WDYT: @findepi @martint

Reply:
@sopel39  I'd prefer pull-based approach to stats derivation. For example, if this is just `SELECT ... WHERE ...` we need no stats during planning.
Also, a pushed down predicate may be refined several times before we ask for stats.
Having noted that, I didn't look into what are the technically viable options yet.

Reply:
> Also, a pushed down predicate may be refined several times before we ask for stats.

That might happen independently if there is `unestimatedPredicate` or not. We always pull stats from connector when needed.

We might have similar issue with projection pushdown. `unestimatedProjection` might be harder to implement though.

Reply:
I have another idea.  Some connectors accept pushed down filter but don't remove it from the plan (i think JDBC connectors do this).
This approach also solves stats problem -- a connector can report stats for the whole table, and the planner will infer stats for the filter predicate, because its still in the plan.

Reply:
> Some connectors accept pushed down filter but don't remove it from the plan (i think JDBC connectors do this).

We can definitely use this approach, but wouldn't it be have a small performance cost (due to the need to pass the scanned rows through the filter operator - that would act as a no-op)?

The motivation behind this PR is to use similar approach to `ConnectorMetadata#applyFilter` API - which returns the "unapplied" predicates to the engine via `ConstraintApplicationResult`.
Since I didn't want to break the existing API - I've added the "unestimated" predicates to the existing `TableStatistics` object.

Reply:
> wouldn't it be have a small performance cost (due to the need to pass the scanned rows through the filter operator - that would act as a no-op)?

It depends on how expensive expressions you're able to pushdown.  How much % can we gain or lose here?

Reply:
A potentially bigger problem with not removing the filter is that it can block further optimizations. For things like JDBC, the cost of Presto applying the filter is insignificant compared to the cost of fetching the data.

(And we should work to support removing the filter in from JDBC.)

Reply:
> How much % can we gain or lose here?

I'll run a performance test, and report back with the result.

Reply:
I ran a [SELECT query with a large filter](https://gist.github.com/rzeyde-varada/b5ca61de727b4794214b2f259c974387#file-query-sql) over TPCDS sf1000 `store_sales` table with predicate pushdown enabled - [once without the FilterNode](https://gist.github.com/rzeyde-varada/b5ca61de727b4794214b2f259c974387#file-without-filter-node-txt), and [once with it](https://gist.github.com/rzeyde-varada/b5ca61de727b4794214b2f259c974387#file-with-filter-node-txt) (similarly to how it's done by the JDBC connector).
IIUC, keeping the FilterNode in the plan increases the CPU time from [~29s](https://gist.github.com/rzeyde-varada/b5ca61de727b4794214b2f259c974387#file-without-filter-node-txt-L28) to [~39s](https://gist.github.com/rzeyde-varada/b5ca61de727b4794214b2f259c974387#file-with-filter-node-txt-L28), and the total elapsed time also increases from [1s](https://gist.github.com/rzeyde-varada/b5ca61de727b4794214b2f259c974387#file-without-filter-node-txt-L41) to [2.5s](https://gist.github.com/rzeyde-varada/b5ca61de727b4794214b2f259c974387#file-with-filter-node-txt-L41) for this specific query.

Reply:
> total elapsed time also increases from 1s to 2.5s for this specific query.

Is Varada data store able to execute "wide OR" much faster than Presto? How do you do this?

Reply:
Btw, what times do you see when you replace this "wide OR" with equivalent IN expression?
@Praveen2112 is working on an optimization (related to https://github.com/prestosql/presto/issues/932) that may potentially roll such OR into an IN, which should be faster to execute.

Reply:
> Is Varada data store able to execute "wide OR" much faster than Presto? How do you do this?

We have an optimized storage engine implementation, supporting highly effective predicate pushdown for most of the filter expressions.

> Btw, what times do you see when you replace this "wide OR" with equivalent IN expression?

Good idea, will check and update with the results.

Reply:
> that may potentially roll such OR into an IN, which should be faster to execute.
If the above predicate is pushed down to the `TableScan` then it happens by default. (provided the connector supports predicate push down). The optimization that we are working will optimize in the case described (#932) or if the connector doesn't support predicate push down.

Reply:
> Good idea, will check and update with the results.

If I use `IN` (instead of wide `OR`), the query finishes after ~1.9 seconds and uses ~37 CPU-seconds (see [here](https://gist.github.com/rzeyde-varada/b5ca61de727b4794214b2f259c974387#file-replace-or-with-in-txt)).